lib/js/jsfetch.nim

```diff
@@ -0,0 +1,153 @@
+
+func values*(this: Headers): seq[cstring] {.importjs: "Array.from(#.values())".}
+  ## https://developer.mozilla.org/en-US/docs/Web/API/Headers/values
+
+func entries*(this: Headers): seq[array[2, cstring]] {.importjs: "Array.from(#.entries())".}
+  ## https://developer.mozilla.org/en-US/docs/Web/API/Headers/entries
+
+template fetchMethodToCstring(metod: HttpMethod): cstring =
+  ## Template that takes an `HttpMethod` and returns an *Uppercase* `cstring`,
+  ## but *only* for the HTTP Methods that are supported by the fetch API.
+  ## High performance and minimal code compared to just `$(HttpMethod)`.
+  assert metod notin {HttpTrace, HttpOptions, HttpConnect}, "HTTP Method not supported by Fetch API"
+  case metod
+  of HttpHead:   "HEAD".cstring
+  of HttpGet:    "GET".cstring
+  of HttpPost:   "POST".cstring
+  of HttpPut:    "PUT".cstring
+  of HttpDelete: "DELETE".cstring
+  of HttpPatch:  "PATCH".cstring
+  else:          "GET".cstring
+
+func hasFetch*(): bool {.importjs: "(() => { return !!window.fetch })()".}
+  ## Convenience func to detect Fetch API support, returns `true` if Fetch is supported.
+
+func unsafeNewFetchOptions*(metod, body, mode, credentials, cache, referrerPolicy: cstring,
+    keepalive: bool, redirect = "follow".cstring, referrer = "client".cstring, integrity = "".cstring): FetchOptions {.importjs:
+    "{method: #, body: #, mode: #, credentials: #, cache: #, referrerPolicy: #, keepalive: #, redirect: #, referrer: #, integrity: #}".}
+  ## **Unsafe** `newfetchOptions`. Low-level func for optimization.
+
+func newfetchOptions*(metod: HttpMethod, body: cstring,
+    mode: FetchModes, credentials: FetchCredentials, cache: FetchCaches, referrerPolicy: FetchReferrerPolicies,
+    keepalive: bool, redirect = frFollow, referrer = "client".cstring, integrity = "".cstring): FetchOptions =
+  ## Constructor for `FetchOptions`.
+  result = FetchOptions(metod: fetchMethodTocstring(metod), body: body, mode: $mode,
+    credentials: $credentials, cache: $cache, referrerPolicy: $referrerPolicy,
+    keepalive: keepalive, redirect: $redirect , referrer: referrer, integrity: integrity)
+
+func fetchToCstring*(url: cstring): cstring {.importjs: "await fetch(#).then(response => response.text()).then(text => text)".}
+  ## Convenience func for `fetch()` API that returns a `cstring` directly.
+
+func fetchToCstring*(url: cstring, options: FetchOptions): cstring {.importjs: "await fetch(#, #).then(response => response.text()).then(text => text)".}
+  ## Convenience func for `fetch()` API that returns a `cstring` directly.
+
+func fetch*(url: cstring): Response {.importjs: "await fetch(#).then(response => response)".}
+  ## `fetch()` API, simple `GET` only, returns a `Response`.
+
+func fetch*(url: cstring, options: FetchOptions): Response {.importjs: "await fetch(#, #).then(response => response)".}
+  ## `fetch()` API that takes a `FetchOptions`, returns a `Response`.
+
+
+runnableExamples:
+  when defined(nimJsFetchTests):
```

Comment:
case in point: when instead of if prevents this from catching CT errors, like `unsafeNewFetchOption` typo below (should be unsafeNewFetchOptions).

After fixing this it gives:
```nim
rawEcho(cstrToNimstr(await fetch("http://httpbin.org/post", options_520093947).then(response => response.text()).then(text => text)));
                       ^^^^^

SyntaxError: missing ) after argument list
```

in nodejs (via `nim doc -b:js -r lib/js/jsfetch.nim`)

It works with this modification:
```nim
runnableExamples:
  proc fn =
     block:
       ...
```

(since await can only be used inside a function in js)